packages/next/src/client/page-bootstrap.ts

```diff
@@ -1,85 +1,98 @@
 import { hydrate, router } from './'
 import initOnDemandEntries from './dev/on-demand-entries-client'
-import initializeBuildWatcher from './dev/dev-build-watcher'
+import initializeBuildWatcher, {
+  ShowHideHandler,
+} from './dev/dev-build-watcher'
 import { displayContent } from './dev/fouc'
 import { connectHMR, addMessageListener } from './dev/error-overlay/websocket'
 import {
   assign,
   urlQueryToSearchParams,
 } from '../shared/lib/router/utils/querystring'
+import { HMR_ACTIONS_SENT_TO_BROWSER } from '../server/dev/hot-reloader-types'
 
 export function pageBootrap(assetPrefix: string) {
   connectHMR({ assetPrefix, path: '/_next/webpack-hmr' })
 
   return hydrate({ beforeRender: displayContent }).then(() => {
     initOnDemandEntries()
 
-    let buildIndicatorHandler: (obj: Record<string, any>) => void = () => {}
+    let buildIndicatorHandler: ShowHideHandler | undefined
 
-    function devPagesHmrListener(payload: any) {
-      if (payload.action === 'serverError') {
-        const { stack, message } = JSON.parse(payload.errorJSON)
-        const error = new Error(message)
-        error.stack = stack
-        throw error
-      } else if (payload.action === 'reloadPage') {
-        window.location.reload()
-      } else if (payload.action === 'devPagesManifestUpdate') {
-        fetch(`${assetPrefix}/_next/static/development/_devPagesManifest.json`)
-          .then((res) => res.json())
-          .then((manifest) => {
-            window.__DEV_PAGES_MANIFEST = manifest
-          })
-          .catch((err) => {
-            console.log(`Failed to fetch devPagesManifest`, err)
-          })
-      } else if (payload.event === 'middlewareChanges') {
-        return window.location.reload()
-      } else if (payload.event === 'serverOnlyChanges') {
-        const { pages } = payload
+    if (process.env.__NEXT_BUILD_INDICATOR) {
+      initializeBuildWatcher((handler) => {
+        buildIndicatorHandler = handler
+      }, process.env.__NEXT_BUILD_INDICATOR_POSITION)
+    }
 
-        // Make sure to reload when the dev-overlay is showing for an
-        // API route
-        if (pages.includes(router.query.__NEXT_PAGE)) {
-          return window.location.reload()
+    addMessageListener((payload) => {
+      if ('action' in payload) {
```

Comment:
This is the same code as before, the diff is a bit messed up but had to split it by `'action' in payload` in order to do type narrowing.